addons/mass_mailing/models/mail_notification.py

```diff
@@ -0,0 +1,124 @@
+                if 'notification_type' not in values:
+                    values['notification_type'] = 'email'
+                if 'mail_id' in values:
+                    values['mail_id_int'] = values['mail_id']
+                values['campaign_id'] = int(self.env['mailing.mailing'].browse(values.get('mass_mailing_id')).campaign_id)
+                status_info = self._update_trace_status(values)
+                if status_info:
+                    values['notification_status'] = status_info
+                    values['status_update'] = fields.Datetime.now()
+                else:
+                    values['notification_status'] = 'outgoing'
+        return super(MailNotification, self).create(values_list)
+
+    def write(self, values):
+        status_info = self._update_trace_status(values)
+        if status_info:
+            values['notification_status'] = status_info
+            values['status_update'] = fields.Datetime.now()
+        return super(MailNotification, self).write(values)
+
+    def _get_records(self, mail_ids=None, mail_message_ids=None, domain=None):
+        if not self.ids and mail_ids:
+            base_domain = [('mail_id_int', 'in', mail_ids)]
+        elif not self.ids and mail_message_ids:
+            base_domain = [('message_id', 'in', mail_message_ids)]
+        else:
+            base_domain = [('id', 'in', self.ids)]
+        if domain:
+            base_domain = ['&'] + domain + base_domain
+        return self.search(base_domain)
+
+    def set_opened(self, mail_ids=None, mail_message_ids=None):
+        notifs = self._get_records(mail_ids, mail_message_ids, [('opened', '=', False)])
+        notifs.write({'opened': fields.Datetime.now(), 'bounced': False})
+        return notifs
+
+    def set_clicked(self, mail_ids=None, mail_message_ids=None):
+        notifs = self._get_records(mail_ids, mail_message_ids, [('clicked', '=', False)])
+        notifs.write({'clicked': fields.Datetime.now()})
+        return notifs
+
+    def set_replied(self, mail_ids=None, mail_message_ids=None):
+        notifs = self._get_records(mail_ids, mail_message_ids, [('replied', '=', False)])
+        notifs.write({'replied': fields.Datetime.now()})
+        return notifs
+
+    def set_bounced(self, mail_ids=None, mail_message_ids=None):
+        notifs = self._get_records(mail_ids, mail_message_ids, [('bounced', '=', False), ('opened', '=', False)])
+        notifs.write({'bounced': fields.Datetime.now()})
+        return notifs
+
+    @api.model
+    def _gc_notifications(self, max_age_days=180):
+        domain = [
+            ('mass_mailing_id', '=', False),
+            ('is_read', '=', True),
+            ('read_date', '<', fields.Datetime.now() - relativedelta(days=max_age_days)),
+            ('res_partner_id.partner_share', '=', False),
+            ('notification_status', 'in', ('sent', 'canceled'))
+        ]
+        return self.search(domain).unlink()
```

Comment:
@tde-banana-odoo Could we replace this by a regular `power_on` of `ir.autovacuum` ?

Reply:
ping @tde-banana-odoo